webapp/sources/rudder/rudder-rest/src/test/scala/com/normation/rudder/RuleTests.scala

```diff
@@ -0,0 +1,132 @@
+* Public License version 3, when you create a Related Module, this
+* Related Module is not considered as a part of the work and may be
+* distributed under the license agreement of your choice.
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder
+
+import com.normation.rudder.domain.policies.Rule
+import com.normation.rudder.domain.policies.RuleId
+import com.normation.rudder.domain.policies.RuleUid
+import com.normation.rudder.rest.RestTest
+import com.normation.rudder.rest.RestTestSetUp
+import com.normation.rudder.rule.category.RuleCategory
+import com.normation.rudder.rule.category.RuleCategoryId
+import net.liftweb.common.Loggable
+import org.junit.runner.RunWith
+import org.specs2.mutable.Specification
+import org.specs2.runner.JUnitRunner
+
+@RunWith(classOf[JUnitRunner])
+class RuleTests extends Specification with Loggable {
+
+  val restTestSetUp = RestTestSetUp.newEnv
+  val restTest = new RestTest(restTestSetUp.liftRules)
+
+  val subCat =
+    List (
+      RuleCategory(RuleCategoryId("cat1"), "", "", List.empty)
+    , RuleCategory(
+        RuleCategoryId("cat2")
+      , ""
+      , ""
+      , List (
+          RuleCategory(RuleCategoryId("cat3"), "", "", List.empty)
+        , RuleCategory(RuleCategoryId("cat4")
+          , ""
+          , ""
+          , List(
+              RuleCategory(
+                RuleCategoryId("subcat4")
+              , ""
+              , ""
+              , List(
+                  RuleCategory(RuleCategoryId("subsubcat4"), "", "", List.empty)
+                )
+              )
+          )
+        )
+      )
+    )
+  )
+
+  val root = RuleCategory(
+    RuleCategoryId("rootRuleCategory") //root category of all categories, it should not be missing
+  , "Root category"
+  , "base root category"
+  , subCat
+  , true
+  )
+
+  "Testing rule utility tools" should {
+    "List all categories and subcategories" in {
+      val listCatIds = Set(
+          RuleCategoryId("rootRuleCategory")
+        , RuleCategoryId("cat1")
+        , RuleCategoryId("cat2")
+        , RuleCategoryId("cat3")
+        , RuleCategoryId("cat4")
+        , RuleCategoryId("subcat4")
+        , RuleCategoryId("subsubcat4")
+      )
+      restTestSetUp.ruleApiService14.listCategoriesId(root) shouldEqual(listCatIds)
+    }
+
```

Comment:
you should add a test that you only get cat4 and children if starting from cat4 (locality) and that you correctly get nothing if non existing cat and that you only get one cat starting from cat1